src/CTA.Rules.Actions/ActionHelpers/ServerConfigMigrate.cs

```diff
@@ -0,0 +1,193 @@
+        }
+        public void Init()
+        {
+            _requiredDirectives = new List<string>();
+            _middleWareExpressions = new Dictionary<string, List<string>>();
+            foreach (var methodName in ServerConfigTemplates.middlewareConfigMethods)
+            {
+                _middleWareExpressions.Add(methodName, new List<string>());
+            }
+
+            _startupFilePath = Path.Combine(_projectDir, FileTypeCreation.Startup.ToString() + ".cs");
+            var code = File.ReadAllText(_startupFilePath);
+            root = CSharpSyntaxTree.ParseText(code).GetRoot();
+        }
+
+        public void ParseServerConfigXML(ConfigurationSection serverConfig)
+        {
+            XDocument xDocument = XDocument.Parse(serverConfig.SectionInformation.GetRawXml());
+            var element = xDocument.Elements().ToList();
+
+            //Handle Http modules
+            var httpModules = element.FirstOrDefault().Descendants(WebServerConfigAttributes.Modules.ToString().ToLower());
+            if (httpModules.HasAny())
+            {
+                HandleHttpModules(httpModules);
+            }
+
+            //Handle Http handlers
+            var httpHandlers = element.FirstOrDefault().Descendants(WebServerConfigAttributes.Handlers.ToString().ToLower());
+            if (httpHandlers.HasAny())
+            {
+                HandleHttpHandlers(httpHandlers);
+            }
+
+            // Handle security Attributes
+            var securityAttribute = element.FirstOrDefault().Descendants(WebServerConfigAttributes.Security.ToString().ToLower());
+            if (securityAttribute.HasAny())
+            {
+                HandleAuthorization();
+
+                //Handle authentication
+                var authenticationAttribute = securityAttribute.FirstOrDefault().Descendants(WebServerConfigAttributes.Authentication.ToString().ToLower());
+                if (authenticationAttribute.HasAny())
+                {
+                    HandleAuthentication(authenticationAttribute);
+                }
+
+            }
+
+            RegisterMiddlewareComponents();
+
+            File.WriteAllText(_startupFilePath, root.ToFullString());
+        }
+
+        // Handle web server attributes
+        private void HandleHttpModules(IEnumerable<XElement> httpModules)
+        {
+            foreach (var module in httpModules.Elements())
+            {
+                // only inject added modules to middleware
+                if (module.Name.ToString().Trim().Equals("add"))
+                {
+                    var moduleName = GetAttributeValue(module, "type", ".");
+                    _middleWareExpressions[ServerConfigTemplates.ConfigureMiddlewareMethod].Add(string.Format(ServerConfigTemplates.configurationExpressionTemplates[WebServerConfigAttributes.Modules.ToString()], moduleName));
+                }
+            }
+        }
+
+        private void HandleHttpHandlers(IEnumerable<XElement> httpHandlers)
```

Comment:
could you create a .net project with simple HttpHandler and after porting could you see your code is doing the job ? I am bit concern the way type being extracted, eg:- if it stays in a different library like this type="BirojHandle.SaiClass, PACustomHandler" .... I would suggest to get the whole thing and give a compilation error for user to go and fix....

Reply:
All the code changes made here are validated against a sample application